Feature form builder

buchholz-sean opened the pull request:
Created dynamic HTML form builder feature. Takes Object input and converts to HTML form elements.

buchholz-sean committed:
Create dynamic html form builder tool

buchholz-sean committed:
Update README.md

buchholz-sean committed:
Create demo file with example content

buchholz-sean committed:
Update for legibility

buchholz-sean committed:
- Added more supported input types: Password, Reset, & Submit
- Added support for additional optional properties per input type (see README)

buchholz-sean committed:
Clarified section detailing supported input object properties

buchholz-sean committed:
Simplified string building within switch case expression

joegreen2 commented on src/tool.js:
```diff
@@ -0,0 +1,57 @@
+                        break;
+                    default:
+                        // If input doesn't match a supported value for `type`, we don't add anything to formStr
+                        formStr += ''
```
Would suggest adding a semicolon at end of line 48.
``` formStr += ''; ```
[ES5 semicolons](http://es5.github.io/#x7.9)
[ES6 semicolons](http://www.ecma-international.org/ecma-262/6.0/index.html#sec-automatic-semicolon-insertion)

(I know this comment is incredibly picky but your code is really solid, so far I can't find anything else)

cayouetten replied:
My only thought other than ES6 implementation would be to deliver feedback to the user here describing which submitted type could not be used, such as

> `console.log(input.type, 'type not supported.');`

buchholz-sean replied:
Nice catch! Fixed.

Rhuansantos commented on src/tool.js:
```diff
@@ -0,0 +1,39 @@
+exports.genForm = (inputs) => {
+    // Create a blank string to start with
+    var formStr = '';
```
You could change `var` to `let` or `const` just to keep the consistency of your code!

buchholz-sean replied:
Good call. Changed `var` to `let` to maintain consistency and avoid potential global var collisions.

Rhuansantos commented on src/tool.js:
```diff
@@ -0,0 +1,39 @@
+    try {
+        inputs.forEach((input) => {
+            // Make sure input has property `type`
+            if (input.type) {
```
Instead of having if and switch statement you could just have an array with valid input types and one if statement to check if the values are in the array. It could save you a lot of lines!

example:
```js
let validInputs = ['text', 'password', 'label', 'select', 'reset', 'submit'];

if(validInputs.includes(input.type)){

}
```

buchholz-sean replied:
This comment seems to address the same concern as your comment below, so I'm only responding to one.

I understand the approach you're trying to take here, and I appreciate opportunities to eliminate lines of code.

That said, I'd prefer to keep the `switch` expression over your pattern for the following reasons:

- Some HTML form elements are not `input` tags and are not self-closing (`select` and `label` are two examples)
- Not all HTML attributes are supported by each type of input or form element
- Future scalability might include things like adding an `option` property to the `select` type with an array of objects, which doesn't work with this one-size-fits-all approach

To elaborate on each point: `select` and `label` elements both require closing tags (along with `textarea` and a handful of others not currently supported by this tool. A pattern like

```
if(validInputs.includes(input.type)){
  returnString=`<input type="${input.type}" ... />`;
}
```

can't accommodate something like `select`.

Not all of the HTML form elements take the same attributes. If the end goal is to produce valid, semantic, syntactically-compliant HTML forms dynamically, there is no benefit to passing, for instance, a `for` attribute to anything except `label` elements. To account for this, we'd need to provide some conditionality anyway, and we wind up with a stack of `if...else` statements. For all the different combinations of supported attributes, we'd need quite a few of them, which isn't great:

>In JavaScript, `if` statements are generally faster than `switch` statements when there are just _one or two conditions_ to be evaluated. When there are more than two conditions, and the conditions are simple (not ranges), the `switch` statement tends to be faster.

(Read more about efficiency in Javascript [here](http://archive.oreilly.com/pub/a/server-administration/excerpts/even-faster-websites/writing-efficient-javascript.html)).

Finally, the pattern above can't accommodate nested elements, like `option`s inside a `select`, or a bunch of radio buttons inside a `fieldset`.

In this case, I believe the `switch` expression to be a much simpler and more elegant solution, despite adding weight to the code.

Rhuansantos replied:
I agree on that! So probably another suggestion could be combining both options and use the switch statement according to Html tag and not the input!

Rhuansantos commented on src/tool.js:
```diff
@@ -0,0 +1,39 @@
+            if (input.type) {
+                // Switch expression handles accepted input `type` props.
+                // toLowerCase makes sure we don't have to handle case sensitivity
+                switch (input.type.toLowerCase()) {
```
I don't think you need the switch statement here imagine if you want to validate all of the possibilities.
Also, you could use template strings to make your code cleaner!

Here is an example of the solution that thought to eliminate repetitive code:
you could have an array and push the inputs for that array if it is valid.

`let fields = []`

Example:
```js
  formStr = `<input type="${input.type}" name="${input.name}" id=${input.id}/>`;
  fields.push(formStr);
```

the <`input`> could also be a template string!

Rhuansantos reviewed (commented):
Hi Sean Buchhol, here are some of my considerations for your code!

cayouetten reviewed (commented):
This is mostly geared to user validation going forward - not so much for this assignment. Overall, concise and relevant comments, great organization and readability of the string build, and good idea with the toLowerCase addition for case insensitivity.

buchholz-sean committed:
Address issues in Pull Request

-  Added table to `README.md` to better demonstrate supported input types and attributes (Per @reactivepixel via Slack, not in PR)
- Better support for various HTML attributes in `tool.js`
- Replaced concatenation with string templates in `tool.js`
- Added more attributes & examples to `demo.js`

buchholz-sean commented:
Please see latest commit for fixes.

buchholz-sean committed:
Fix minor table formatting issue

buchholz-sean committed:
Fix inconsistent grammar in How It Works section

buchholz-sean committed:
Consolidate cases in `switch`

Combined `email`, `password`, and `text` cases and adjusted start of `formStr` to dynamically create `input` elements of those `type`s

cayouetten reviewed (approved):
Great job, Sean

Rhuansantos reviewed (approved):
Awesome job combining the switch statements!
